### ISSUE TITLE
Add support for PBMAC1 [RFC 8018 section 7.1]

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#20072]
PBMAC1 is  defined in [RFC 8018 section 7.1](https://www.rfc-editor.org/rfc/rfc8018.html#section-7.1.), obsoleting [RFC 2898](https://www.rfc-editor.org/rfc/rfc2898).
Interestingly, an NID is defined in OpenSSL since 2015, but no actual implementation is provided.
It is supported by other crpyto libraries such as Bouncy Castle and by and by the EJBCA of Keyfactor.

Would be nice if this was available, at least for use in CMP.
Password-based MAC (PBM) is supported as per RFC 4210/4211, but is outdated.
[CMP Algorithms RFC 9481 suggests](https://www.rfc-editor.org/rfc/rfc9481.html#name-pbmac1) using instead the more secure PBMAC1.
This is already supported by Bouncy Castle 

PBMAC1 is needed by CMP as a replacement for the outdated [Password-Based MAC (PBM)](https://www.rfc-editor.org/rfc/rfc4211.html#section-4.4), for:
* MAC-based message integrity protection https://www.rfc-editor.org/rfc/rfc9481.html#name-pbmac1
* POP processing: https://www.rfc-editor.org/rfc/rfc9045.html#name-signature-key-pop

Update: looks like it is also needed for [supporting PKCS#12](https://github.com/openssl/openssl/issues/19997).

**Comment by beldmit** [comment]
I support this because we need it for one of our drafts.

I have a very draft implementation of PBMAC1, can be used as a starting point (or thrown away): https://github.com/beldmit/openssl/tree/pkcs12_pbmac1

**Comment by mattcaswell** [comment]
Also see #19997 where PBMAC1 support is discussed. It's actually a fairly trivial combination of PBKDF2 and a HMAC IIUC. There is some debate in #19997 as to whether we would want to make this a "fetchable" provider based thing or not.

**Comment by DDvO** [comment]
@beldmit meanwhile started implementing https://datatracker.ietf.org/doc/draft-ietf-lamps-pkcs12-pbmac1/ 
at https://github.com/openssl/openssl/compare/master...beldmit:openssl:pkcs12_pbmac1 which contains a preliminary PBMAC1 implementation.

**Comment by DDvO** [comment]
@tomato42 demanded today in https://github.com/openssl/openssl/pull/24577#issuecomment-3091929048:
> on API level it has to be a general API that accepts just password and two objects: KDF params and HMAC params.

Yesterday, I had suggested in https://github.com/openssl/openssl/pull/24577#issuecomment-3089678694
the following default parameter values:
* KDF: PBKDF2
* Salt: randomly generated 20 bytes
* IterationCount: 10000
* KeyLength: 4096
* MacAlgorithm: hmacWithSHA256

**Comment by tomato42** [comment]
@paulidale https://github.com/openssl/openssl/pull/24577#issuecomment-3085741851
and @slontis https://github.com/openssl/openssl/pull/24577#issuecomment-3087174235

no, I really don't see why we would have to put PBMAC1 construction into the fips provider: it's fundamentally the same as PBES2, and we don't have that there. It's just a standardised way to encode PBKDF2 and HMAC parameters into an ASN.1 structure. That's not the kind of code that is in scope for FIPS certification.

and we had that discussion already: https://github.com/openssl/openssl/issues/19997#issuecomment-1373807358 has anything changed?

**Comment by xnox** [comment]
I think this is implemented in 3.4 https://openssl-library.org/news/openssl-3.4-notes/

**Comment by beldmit** [comment]
Unfortunately no, I implemented only a subset back then and not in a universal manner

**Comment by xnox** [comment]
> Unfortunately no, I implemented only a subset back then and not in a universal manner

What are we missing? Using it by default?

**Comment by beldmit** [comment]
I think @DDvO requested an universal recursive ASN.1 parsing. I've implemented the format suitable for RFC 9579 in FIPS mode - not sure it will fit for example if we use scrypt

**Comment by t8m** [comment]
I do not think this RFE is needed. If there is a request for a concrete use-case to be supported with PBMAC1, please open a new issue.
